utils/rowgroup/rowstorage.cpp

```diff
@@ -2267,94 +2305,62 @@ void RowAggStorage::finalize(std::function<void(Row&)> mergeFunc, Row& rowOut)
       else
         prevKeyRowStorage = prevRowStorage.get();
 
-      loadGeneration(prevGen, prevSize, prevMask, prevMaxSize, prevHashMultiplier, prevInfoInc,
-                     prevInfoHashShift, prevInfo);
-      prevHashes = fCurData->fHashes->clone(prevMask + 1, prevGen, true);
-
-      // iterate over current generation rows
-      uint64_t idx{};
-      uint32_t info{};
-      for (;; next(info, idx))
+      if (fExtKeys)
       {
-        nextExisting(info, idx);
-
-        if (idx >= elems)
-        {
-          // done finalizing generation
-          break;
-        }
-
-        const auto& pos = fCurData->fHashes->get(idx);
-        if (fKeysStorage->isFinalized(pos.idx))
-        {
-          // this row was already merged into newer generation, skip it
-          continue;
-        }
-
-        // now try to find row in the previous generation
-        auto [pinfo, pidx] =
-            rowHashToIdx(pos.hash, prevMask, prevHashMultiplier, prevInfoInc, prevInfoHashShift);
+        fKeysRowGroup->initRow(&tmpKeyRow);
+      }
+      auto& prevRow = fExtKeys ? tmpKeyRow : tmpRow;
 
-        while (pinfo < prevInfo[pidx])
-        {
-          ++pidx;
-          pinfo += prevInfoInc;
-        }
-        if (prevInfo[pidx] != pinfo)
+      // iterate over rows in prev generation
+      for (uint64_t rowidx = 0; prevKeyRowStorage->getRowForFinalization(rowidx, prevRow); ++rowidx)
+      {
+        auto [rgid, rid] = rowIdxToGidRid(rowidx, prevKeyRowStorage->getRowsPerRG());
+        if (rgid != 0 && rid == 0)
         {
-          // there is no such row
-          continue;
+          // start next RGData. At this point we don't need the previous one and can free the data
+          // as it has no changes to dump
+          prevKeyRowStorage->dropRGData(rgid - 1);
+          if (fExtKeys) {
+            prevKeyRowStorage->dropRGData(rgid - 1);
+          }
         }
 
-        auto ppos = prevHashes->get(pidx);
-        while (prevInfo[pidx] == pinfo && pos.hash != ppos.hash)
-        {
-          // hashes are not equal => collision, try all matched rows
-          ++pidx;
-          pinfo += prevInfoInc;
-          ppos = prevHashes->get(pidx);
-        }
-        if (prevInfo[pidx] != pinfo || pos.hash != ppos.hash)
+        if (prevKeyRowStorage->isFinalized(rowidx))
         {
-          // no matches
           continue;
         }
 
-        bool found = false;
-        auto& keyRow = fExtKeys ? fKeyRow : rowOut;
-        fKeysStorage->getRow(pos.idx, keyRow);
-        while (!found && prevInfo[pidx] == pinfo)
+        // TODO: store hashes in the RowGroupStorage?
+        uint64_t hash = hashRow(prevRow, fLastKeyCol);
+        auto [info, idx] = rowHashToIdx(hash);
+        nextWhileLess(info, idx);
+        uint64_t curidx = -1;
```

Comment:
Mb worth turning this value into a named const?

Reply:
:+1: